### DISCUSSION TITLE
Making the Cartesian config parser a standalone varianter for Avocado

### DISCUSSION THREAD

**Discussion by pevogam** [avocado-framework/avocado-vt#3772] [Polls]
Hi all, I have often found myself in the need to quickly generate multiple test variants and so far have found the Avocado varianter plugins insufficient in expressiveness and functionality in comparison to the Cartesian configs currently supported only in Avocado VT. If I would like to have more regular non-VT tests (read: tests not needing virtualization) and would like to use the Cartesian configuration, it requires forceful coupling with dependencies and projects. What do you think about the possibility of separating the Cartesian configuration from the Avocado VT project, perhaps as a varianter for Avocado? I won't mind help maintaining such a project and I even have some features planned that could benefit its entire use.

@luckyh, @richtja, @clebergnu, @ldoktor: What do you think about this possibility?

**Comment by richtja** [comment]
Hi @pevogam I am not really familiar with Cartesian varianter in avocado-vt, therefore I am not able to estimate how difficult is to separate the varianter from avocado-vt. From the avocado side, I don't see a problem with having a Cartesian varianter as an optional plugin and It is definitely a welcome contribution. The only thing which concerns me is that it will add new dependence to the avocado-vt, because avocado can be installed without optional plugins.

**Reply by pevogam** [reply]
Hi @richtja, considering that aexpect is a dependency of avocado with small codebase that is quite useful on its own I guess we have already gone that path for one project that is different than avocado and avocado-vt but is under the same umbrella. Of course it would be a second project of that kind and I agree with you in my preference to stay away from dependency overhead as much as possible but I guess one could argue that there are multiple varianter plugins for avocado and this is simply the varianter plugin avocado-vt uses.

**Comment by ldoktor** [comment]
I'm not much involved in Avocado anymore so let me just ask some questions but don't take my opinions for granted. My main question would be, how many users would benefit out of it and whether someone would be prepared to maintain this in the long term (I know your history, Plamen, so I don't doubt the second part).

As for the benefit, I used to be a heavy user of cartesian, it's an amazing tool and if one understands it can really make magic. Similarly with the multiplexer, it was designed to allow almost the same things only in a structured manner. Is there a specific feature you miss in multiplexer that cartesian contains? Maybe it'd be easier to port that feature and use multiplexer instead.

Anyway if you decide you are willing to maintain it (either as a part of Avocado or as a standalone project) and would dedicate that time to create it, then the only problem is the review time of the Avocado team and I can not talk about them.

**Reply by pevogam** [reply]
Hi @ldoktor and thanks for chiming in despite many of these projects mostly being distant echoes of your past. Let me elaborate on your points:

> My main question would be, how many users would benefit out of it and whether someone would be prepared to maintain this in the long term (I know your history, Plamen, so I don't doubt the second part).

I hope at the very least the Avocado VT community will benefit from better integration with Avocado as its code base will be somewhat reduced and further converted into usable plugins. On our side we already don't mind maintaining this repo as our integration test suite benefits from it immensely. Additionally, we plan to use this powerful varianter for the integration tests of further products that have swappable backends and dependencies and can benefit from better test code reusability. The rest is beyond me but hopefully with more documentation, testing, and separation of concerns such a varianter could pick the interest of more people.

> As for the benefit, I used to be a heavy user of cartesian, it's an amazing tool and if one understands it can really make magic. Similarly with the multiplexer, it was designed to allow almost the same things only in a structured manner. Is there a specific feature you miss in multiplexer that cartesian contains? Maybe it'd be easier to port that feature and use multiplexer instead.

I had originally first tried the YAML to Mux plugin during our migration from Autotest to Avocado (here meaning migration away from our large production Cartesian config with hundreds of integration tests) and back then I recall that I encountered quite many missing features prompting me to go the Avocado VT route. Since it was years ago and I cannot recall what the problems were I took a fresh look today and I might be wrong in my refreshing reading of the [documentation about it](https://avocado-framework.readthedocs.io/en/102.0/plugins/optional/varianters/yaml_to_mux.html#yaml-to-mux-plugin) but I see:

1) No support for AND/OR operators on restrictions (e.g. differentiating among `only=fedora..x86_64`, `only=fedora,x86_64` as it can only translate `only=fedora.x86_64` into `!filter-only : /fedora/x86_64`). I assume it still supports restrictions of products (multiplexed subvariants) despite there being no example for it.

2) No support for conditionals, conditional parameters, and conditional restrictions like

```
variants:
    - a:
        param1 = a
    - b:
        param1 = b
b:
    param2=c
    only c (...)
```

3) No support for regex matched parameters like `driver_image1(_vm.*)? ?= virtio`. While appending could be done via lists, prepending or simple replacement using `some_key${prev_key}` could not be done either.

4) No support for suffix and join operators which are much used in our parametric object composition but one viable example for Avocado VT would be having tests with multiple vms of different guest OS and hardware.

> Anyway if you decide you are willing to maintain it (either as a part of Avocado or as a standalone project) and would dedicate that time to create it, then the only problem is the review time of the Avocado team and I can not talk about them.

Indeed I don't mind and I have already created a repo with some initial small improvements. The question is mostly what the Avocado community thinks.

**Comment by luckyh** [comment]
Hi @pevogam I think this is a very interesting proposal! Have you had the draft code already for POC?

**Reply by pevogam** [reply]
Hi @luckyh indeed I do and I even have some small improvements on the documentation and modularity of the plugin already in place. Would you like me to push the repo under my own username or should we create at least a temporary (possibly to delete) repo under the avocado umbrella that you want to see my commits in?

**Reply by pevogam** [reply]
To reduce the back and forth I currently pushed all commits to a PoC repo you might want to check out:

https://github.com/pevogam/cartconf

Let me know what you think of the current commits and the overall idea! Right now one could just import the main package and use all previous functionality of the Cartesian config as well as use (now separated) `parse.py` run script.

My highest priority now is to add more unit/isolation tests in order to explore and validate all functionality, then a second priority would be to refactor and document all of it for readability and reduced maintenance costs, a final still high priority would be to allow arbitrary tests (not just VT tests) to use dictionaries of params produced from such configs.

**Reply by luckyh** [reply]
@pevogam sorry for the late reply. Apart from few minor issues I'd like to open up some discussions the PoC repo looks pretty good to me! However, one thing I'd like to confirm with you is that, as you mentioned in the summary, did you mean by making the cartesian parser an actual varianter for avocado (just like the [varianter_cit](https://github.com/avocado-framework/avocado/tree/master/optional_plugins/varianter_cit))? or merely splitting out the `cartesian_config` module from avocado-vt? Because I didn't see the code about the plugin definition from the PoC repo, I'm not sure if I missed something or that was not a goal.

Anyway, at the moment, I think we could try to create such a standalone repo and make it a submodule of the avocado-vt repo, and install it during the vt's installation. With this approach, I guess the vt repo would not be bothered by the transition of the cartesian conf functionalities too much.

**Reply by luckyh** [reply]
And regarding the minor issues:

1. suggest dropping the ill-defined features for the standalone repo. For example [this one](https://github.com/avocado-framework/avocado-vt/blob/522713f3593988e728046002378f108c94942520/virttest/cartesian_config.py#L2265C5-L2265C12), from my perspective it should belong to `utils_param` instead of being a part of the syntax.
2. currently the [reserved keys](https://github.com/pevogam/cartconf/blob/1900be6b546d8f79b1b521bfbdb792e18a42fafc/cartconf/constants.py#L5C1-L5C92) was restored together with the user defined keys, which requires the users to be learnt and aware on that. For the standalone repo, I think now it's a good timing to improve that by introducing a dedicated dict (e.g. `metadata`) to separate them.
3. suggest renaming the [`parse`](https://github.com/pevogam/cartconf/blob/main/cartconf/parse.py) module to `cli` and make it to be imported by `__init__` in favor of user then could run it by `python -m cartconf`.

**Reply by pevogam** [reply]
> However, one thing I'd like to confirm with you is that, as you mentioned in the summary, did you mean by making the cartesian parser an actual varianter for avocado (just like the [varianter_cit](https://github.com/avocado-framework/avocado/tree/master/optional_plugins/varianter_cit))? or merely splitting out the `cartesian_config` module from avocado-vt? Because I didn't see the code about the plugin definition from the PoC repo, I'm not sure if I missed something or that was not a goal.

Indeed, this is exactly the goal. As the above was just an initial PoC on how a standalone project might be structured and might look like (and I wasn't sure if there is even any interest in pushing forward with it), I didn't take the steps yet to provide the varianter plugin interface. I am definitely eager to look into this next as long as such a proposal gets approved and we set up an official project.

> Anyway, at the moment, I think we could try to create such a standalone repo and make it a submodule of the avocado-vt repo, and install it during the vt's installation. With this approach, I guess the vt repo would not be bothered by the transition of the cartesian conf functionalities too much.

We could do this until the inferface to a varianter plugin is provided but once it becomes available I think it should be just as easy to set up Avocado VT with the Cartesian config as setting up Avocado with two plugins, namely the former two.

> And regarding the minor issues:

All great suggestions! I can register each one of them as an initial issue to deal with (+ one issue for the varianter plugin interface) once we set up a project. Then at first opportunity I will look into these issues in order of priority.

**Reply by pevogam** [reply]
@luckyh Have you had time to consider the example project? I will prioritize more of your points if you can confirm we are all interested and perhaps provide me with a location to establish the repo.

**Reply by luckyh** [reply]
> @luckyh Have you had time to consider the example project? I will prioritize more of your points if you can confirm we are all interested and perhaps provide me with a location to establish the repo.

@pevogam sorry I just got lost in other tasks and almost forgot this one, thanks for the reminder! I think there are merely a few things to be addressed before we are going to establish the repo.

The first thing is, @clebergnu @richtja I would like to hear your opinions about if you would like to have the new repo placed under the avocado umbrella, or if you would have some other suggestions on that. Thanks in advance!

**Reply by richtja** [reply]
Hi @luckyh and @pevogam, from my point of view there is no need to separate this into its own repo. IMO this could be well maintained as avocado optional plugin, the same way as [CIT Varianter](https://github.com/avocado-framework/avocado/tree/master/optional_plugins/varianter_cit), [PICT Varianter](https://github.com/avocado-framework/avocado/tree/master/optional_plugins/varianter_pict) or [YAML to Mux Varianter](https://github.com/avocado-framework/avocado/tree/master/optional_plugins/varianter_yaml_to_mux).

**Reply by pevogam** [reply]
Hi @richtja my point is exactly to have it as such an optional avocado plugin and thus to be used for variant generation beyond Avocado VT and its VT tests. If it remains part of the Avocado VT repo it remains implicitly coupled with VT tests and everything else the repo contains and in order to release a new version of the optional plugin we will have to release a new version of Avocado VT which is too much coupling IMO. What are your thoughts about this?

**Reply by richtja** [reply]
Yes, I understand your reasons for moving the Cartesian config parser from avocado-vt repo, and I agree with it. My proposal is about moving it into avocado repo as optional plugin instead of creating new repo. With such change, we will provide the Cartesian config parser to the other test types without creating and maintaining CI infrastructure for new repo. The only disadvantage is that it will be tied up to the avocado release as the other optional plugins. 

By the way, this doesn't have to be the final solution, because there is a plan to move optional plugins into its own repo in https://github.com/avocado-framework/avocado/issues/4389. Right now, we are working on infrastructure for moving avocado's and avocado-vt's utils to https://github.com/avocado-framework/autils and part of that work are https://github.com/avocado-framework/avocado-static-checks and https://github.com/avocado-framework/avocado-ci-tools projects for sharing CI standards between repos. When the migration of autils will be done, we can move to optional plugins and use the same concepts.

**Reply by pevogam** [reply]
I see, the immediate problem I can foresee with this is that I want to do a lot of intense development on the plugin to add various unit/isolation tests, refactor and document current functionality, and possibly add new features like faster processing, complete join and suffix operators, and so on. All of this means that I will have to do a lot of commits and merge a lot of branches and at the current rate of pull request merges in avocado this will be really limiting for me and introduce bottlenecks for maintenance. I already maintain one plugin for Avocado (Avocado I2N) so I don't main maintaining another one and I am afraid while easier at first, keeping this under the avocado umbrella will create a lot of extra work for everyone involved. If it is a separate repo I will be able to push and release more often and maintain compatibility with Avocado VT and other potential user code on stable releases.

**Reply by richtja** [reply]
To be hones, I don't see how developing it in different repo under avocado organization will increase the rate of pull request merges. IMO, the PR review process will have to be the same as it is in avocado to preserve the same quality. The Avocado I2N is different story because it is not under Avocado organization.

**Reply by pevogam** [reply]
I guess I wasn't clear - I want to actively work on this plugin at least for a while and perform lots of changes/updates and I need to be able to push and deploy frequently and this is not just a suggestion for moving repo and just maintaining the current code base. This is not about reducing the quality or avoiding scrutiny and pull request / review workflow. In the past years it could take me anywhere from a few weeks to more than a year to merge a single pull request with avocado so this is not something that is compatible with the rather long list of rather big changes I am planning. But we might as well have a separate repo for the plugin under a different organization, I just thought it would be better if everything is more integrated under the avocado umbrella.

**Reply by richtja** [reply]
I understand that merging PR to Avocado, especially with the lack of reviewers, can be time-consuming and create bottlenecks in your work and IMO this is the biggest struggle of upstream development. But I just still don't understand how moving the work to other repo will help if you don't want to remove pull request / review workflow. What could help, would be to bring more people to this (minimally, for reviews), which I think is possible since this has a common goal for avocado and avocado-vt.

Of course, if you want to do pushes without PRs and have more control over the project, that is a different debate then and IIANM we don't have such repo in avocado organization for now. Therefore, it will be a much bigger organization change.

There is also a possibility to do most of the main changes in your own repo and move it to avocado for maintenance when it will be done.

**Reply by pevogam** [reply]
> I understand that merging PR to Avocado, especially with the lack of reviewers, can be time-consuming and create bottlenecks in your work and IMO this is the biggest struggle of upstream development. But I just still don't understand how moving the work to other repo will help if you don't want to remove pull request / review workflow. What could help, would be to bring more people to this (minimally, for reviews), which I think is possible since this has a common goal for avocado and avocado-vt.

I agree that bringing more people to the boat is the crux of the problem, however unfortunately it remains an unsolved problem that would block development on the Cartesian config if we accept it as a counterargument. Then indeed what I mean is that if there are no reviewers in due time and all tests are passing (of which I plan to be more) I would tend to merge a pull request and block further development.

> Of course, if you want to do pushes without PRs and have more control over the project, that is a different debate then and IIANM we don't have such repo in avocado organization for now. Therefore, it will be a much bigger organization change.

A separate repo within avocado must still be possible where I could be give the task of maintain and freedom to push more often and this is not something that could happen under the core avocado repo which was your originally suggested alternative.

> There is also a possibility to do most of the main changes in your own repo and move it to avocado for maintenance when it will be done.

If I commit a lot of changes in a separate repo to later move for maintenance under avocado then I would also prefer (and I think it is most preferable for everyone) if I am able to maintain it also afterwards, especially considering a likely large number of changes introduced by me. So even in this case a separate repo would be better which results in next to no difference with the original proposal.

So I guess the question boils down to: separate avocado repo under the same or a different umbrella. Or did I miss something in your already quite good suggestions?

**Reply by richtja** [reply]
Yes, basically you are right. After what you said, I think the whole question is about approval policies. In Avocado, every change needs to have at lease one approval. So if we decide that we won't enforce this, then we can have this repo under the avocado umbrella. If we keep enforcing it, then I understand that you will move it somewhere else because it will slow you down.

 I think it would be great to get opinion about this also from @luckyh, @clebergnu, @ldoktor.

**Reply by luckyh** [reply]
I would agree with @richtja that it'd better to comply the same review workflow as what we've set for other avocado repos, if we want the new repo to be kept in the avocado organization.

In a long-term perspective, there will be not that much new features/functionalities/unittests to be added into the Cartesian package, hence I'd confidently say it won't be a real issue which could significantly slow down the development progress if we choose such a "strict" approval policy.

However, I understood that there would be tens of code/PRs submission at the early stage of the new repo where a delayed review could be even seen as the major stumbling block. On the other hand, I do believe code reviews can help on sustaining the stability, so I had the same thought as @clebergnu commented below. I mean, we can have the only review instead of several ones, which just happens once the new package is ready for GA.

**Comment by pevogam** [comment]
One idea regarding all the module headers: the best approach we could come up with is to add copyright headers and the original authors listed when they were working on this as is done in the linux kernel. An example header from `mm/memcontrol.c`:
```
// SPDX-License-Identifier: GPL-2.0-or-later
/* memcontrol.c - Memory Controller
 *
 * Copyright IBM Corporation, 2007
 * Author Balbir Singh <balbir@linux.vnet.ibm.com>
 *
 * Copyright 2007 OpenVZ SWsoft Inc
 * Author: Pavel Emelianov <xemul@openvz.org>
 *
 * Memory thresholds
 * Copyright (C) 2009 Nokia Corporation
 * Author: Kirill A. Shutemov
 *
 * Kernel Memory Controller
 * Copyright (C) 2012 Parallels Inc. and Google Inc.
 * Authors: Glauber Costa and Suleiman Souhlal
 *
 * Native page reclaim
 * Charge lifetime sanitation
 * Lockless page tracking & accounting
 * Unified hierarchy configuration model
 * Copyright (C) 2015 Red Hat, Inc., Johannes Weiner
 *
 * Per memcg lru locking
 * Copyright (C) 2020 Alibaba, Inc, Alex Shi
 */
```

**Reply by ldoktor** [reply]
Looking at `cartesian_config.py` I can see only `:copyright: Red Hat 2008-2013`. Sure there were other significant changes afterwards by RH as well as others, but I'd simply list there that it is `refactored-from: https://github.com/avocado-framework/avocado-vt` followed by the original copyright + your copyright with the current date.

Anyway I'm not a lawyer nor I favor licenses. But as long as you point to the original authors and keep the license I assume it should be fine.

**Comment by pevogam** [comment]
One additional idea I have once we establish enough many unit tests for acceptance testing is to convert some of the core functionality to Rust with the same python interface as exported python bindings. Any objections or affirmations are welcome and I hope this doesn't sound too far extreme even if it could sound far fetched until acceptance testing becomes possible for such a repo.

**Comment by clebergnu** [comment]
Hi @pevogam,

When it comes to the organization of the repo, it's a fair game to have a new repo under the "avocado-framework" organization, with relaxed code-review guidelines until the project matures to the point you consider it "GA".

Then, I'd recommend one of two things:

1. Make it a "first-class" Avocado plugin.  The fact that the cartesian config doesn't have dependency on external Python libraries allows for that.  If you're concerned about further development of the plugin, like waiting on reviews, you'd be free to count on say, a colleague from your company, to perform these reviews and still satisfy the review criteria.
2. Keep it in the "avocado-framework" organization, as a separate repo, but enforce the coding/review guidelines as the other projects do too.

What do you think?

**Reply by pevogam** [reply]
Hi Cleber! Right now I think option (1) is more flexible to the requirements of such a project. Of course if you happen to have anyone at Red Hat that would like to perform frequent reviews and spend more of their resources in this project, they are always welcome to do so and greatly appreciated. Otherwise I will look for a person within my own team that would hopefully agree to do the same so that we stay as close to Avocado's review principle as possible, even at this stage. Then of course once we reach a point of long term stability we could embrace the full code of conduct and avocado workflows. Does this sound good?

**Comment by pevogam** [comment]
In the meantime while waiting for further feedback and potential action I will push some further work to https://github.com/pevogam/cartconf. I intend to introduce some Rust conversions via acceptance testing there too.

**Reply by pevogam** [reply]
This starts with some exploratory isolation tests that cover all the main smaller classes (label, node, lexer, string and file readers, etc.) that raises the coverage from the original
```
..............
----------------------------------------------------------------------
Ran 14 tests in 0.007s

OK
```
to
```
...............................................
----------------------------------------------------------------------
Ran 47 tests in 0.008s

OK
```
I also added various missing docstrings and type hints for these classes while using the new tests to study the code base (thus writing down all the types and docstrings to be added for faster reference later on). Details are visible at this minimal acceptance testing branch:

https://github.com/pevogam/cartconf/tree/minimal-acceptance-testing

**Reply by pevogam** [reply]
Current acceptance testing stage:
```
.............................................................
----------------------------------------------------------------------
Ran 61 tests in 0.010s

OK
```

**Reply by pevogam** [reply]
This branch has been merged and we now have 3 github worklows on the main branch you can see here:

CI: https://github.com/pevogam/cartconf/actions/runs/13941651551
Review: https://github.com/pevogam/cartconf/actions/runs/13941651562
Lint: https://github.com/pevogam/cartconf/actions/runs/13941651575

**Reply by pevogam** [reply]
Further increase in acceptance testing coverage on the main branch:
```
.............................................................................................................................
----------------------------------------------------------------------
Ran 125 tests in 0.014s

OK
```

**Comment by pevogam** [comment]
A new branch https://github.com/pevogam/cartconf/tree/varianter-plugin now has the possibility to run `cartconf` as a varianter plugin for avocado (see README usage and installation sections there). @richtja @clebergnu This is an optional plugin with more powerful variantization capabilities (e.g. join operations) that if you want we can move under the avocado organization at least? In the longer term I would also recommend letting Avocado VT make full use of it too.

**Reply by pevogam** [reply]
I added integration tests for this branch for both entry points now and have now merged this branch. Feel free to find these and any follow up changes in the main branch.

**Comment by pevogam** [comment]
A Rust-Python mixed code base is introduced via PyO3 in https://github.com/pevogam/cartconf/tree/rust-codebase-mix and the CI for both the python and rust side can be found in https://github.com/pevogam/cartconf/actions/runs/13718182501. I intend to start migrating some code to Rust for speed-up after some more added acceptance tests in the near future.

**Comment by pevogam** [comment]
A general update on the cartconf (Cartesian config) plugin for everyone, right now the rust implementation performs on par with the python implementation when it comes to large production-grade join-inclusive benchmark:

```
# branch main-rs (rust implementation)
time ./parse.py test4.cfg
...
dict 60150:  pyi2ncommon.common.cnfvar_api.arnied_api.mis1.qcow2.virtio_blk.smp2.virtio_net.q35.Intra2net.Business_Server.official.mis2.qcow2.virtio_blk.smp2.virtio_net.q35.Intra2net.Business_Server.official
real    13m5.713s
user    12m41.009s
sys     0m13.320s
```

vs

```
# branch main (python original)
time ./parse.py test4.cfg
...
dict 60150:  pyi2ncommon.common.cnfvar_api.arnied_api.mis1.qcow2.virtio_blk.smp2.virtio_net.q35.Intra2net.Business_Server.official.mis2.qcow2.virtio_blk.smp2.virtio_net.q35.Intra2net.Business_Server.official
real    12m8.898s
user    12m0.739s
sys     0m0.327s
```
For join-free parsing the rust implementation still performs somewhat worse than the python implementation but:

1) We are investigating some ways to provide hybrid keys for the params of the parsed dictionaries.
2) The rust implementation actually provides the possibility to cache the parsed AST and thus significantly reduce re-parsing needs - something that python due to its maximized reuse and thus coupling does not allow.

Both branches are available for you to try as `main` (python) and `main-rs` (rust) respectively. The CI reveals the way you can simply include this as avocado plugin but you can always also call the `parse.py` entry point directly.

**Comment by pevogam** [comment]
To follow up on this with somewhat of a resultion I have created a repo everyone could look into at:

https://github.com/avocado-framework/cartconf

Feel free to provide any feedback there, I hope we can accumulate enough interest on the VT side so that we can define this as a proper dependency for the VT project and drop the entire `cartesian_config.py` file from the VT code base.
